src/alert-manager/src/alert-handler/controllers/mail.js

```diff
@@ -130,73 +59,117 @@ const sendEmailToAdmin = async (req, res) => {
         externalURL: req.body.externalURL,
       },
     })
-    .then(function () {
+    .then(() => {
       logger.info(
         `alert-handler successfully send email to admin at ${process.env.EMAIL_CONFIGS_ADMIN_RECEIVER}`,
       );
       res.status(200).json({
         message: `alert-handler successfully send email to admin at ${process.env.EMAIL_CONFIGS_ADMIN_RECEIVER}`,
       });
     })
-    .catch(function (data) {
-      logger.error('alert-handler failed to send email to admin');
-      logger.error(data);
+    .catch((error) => {
+      logger.error('alert-handler failed to send email to admin', error);
       res.status(500).json({
         message: `alert-handler failed to send email to admin`,
       });
     });
 };
 
+const getUserNameByJobName = (jobName, token) => {
+  return axios
+    .get(`${process.env.REST_SERVER_URI}/api/v2/jobs/${jobName}`, {
+      headers: {
+        Authorization: `Bearer ${token}`,
+        'Content-Type': 'application/json',
+      },
+    })
+    .then((response) => {
+      return response.data.jobStatus.username;
+    });
+};
+
+const getUserEmail = (username, token) => {
+  return axios
+    .get(`${process.env.REST_SERVER_URI}/api/v2/users/${username}`, {
+      headers: {
+        Authorization: `Bearer ${token}`,
+        'Content-Type': 'application/json',
+      },
+    })
+    .then((response) => {
+      return response.data.email;
+    });
+};
+
+// send email to job user
 const sendEmailToUser = async (req, res) => {
-  // send email to job user
-  // group alerts by username
-  const url = process.env.REST_SERVER_URI;
-  const token = req.token;
-  let alertsGrouped;
+  // extract jobs names
+  const alerts = req.body.alerts.filter(
+    (alert) => alert.status === 'firing' && 'job_name' in alert.labels,
+  );
+  if (alerts.length === 0) {
+    return res.status(200).json({
+      message: 'No alert need to be send to users.',
+    });
+  }
 
+  // group alerts by username
+  const alertsGrouped = {};
   try {
-    alertsGrouped = await getAlertsGroupedByUser(req.body.alerts, url, token);
-  } catch (data) {
-    logger.error(data);
-    res.status(500).json({
-      message: `Send email encourted Unknown Error`,
+    await Promise.all(
+      alerts.map(async (alert) => {
+        const userName = await getUserNameByJobName(
+          alert.labels.job_name,
+          req.token,
+        );
+        if (userName && userName in alertsGrouped) {
+          alertsGrouped[userName].push(alert);
```

Comment:
map function will run parrallel, does push function is thread-safe?
Can you get all names first then do some group action?